src/game_updater/PlayerUpdater.cpp

```diff
@@ -0,0 +1,298 @@
+  // use basic key detection when on the ground, newly pressed for doublejump
+  if ((player->GetGround() && ImGui::GetIO().KeysDown[GLFW_KEY_SPACE]) ||
+      (!player->GetGround() &&
+       InputBindings::KeyNewlyPressed(GLFW_KEY_SPACE))) {
+    Jump(game_state);
+  }
+
+  // apply gravity
+  if (!player->GetGround()) {
+    player->SetYVelocity(player->GetYVelocity() - PLAYER_GRAVITY);
+  }
+
+  glm::vec3 prevCameraPos = camera->getPosition();
+  // always check for ducking (be ware of ducks)
+  // let the player move up/down Z (sock it to me)
+  if (ImGui::GetIO().KeysDown[GLFW_KEY_LEFT] ||
+      ImGui::GetIO().KeysDown[GLFW_KEY_H]) {
+    player->MoveDownZ();
+    player->SetDucking(DuckDir::LEFT);
+    camera->setPosition(prevCameraPos -
+                        glm::vec3(0, 0, PLAYER_DELTA_Z_PER_TICK));
+  } else if (ImGui::GetIO().KeysDown[GLFW_KEY_RIGHT] ||
+             ImGui::GetIO().KeysDown[GLFW_KEY_L]) {
+    player->MoveUpZ();
+    player->SetDucking(DuckDir::RIGHT);
+    camera->setPosition(prevCameraPos +
+                        glm::vec3(0, 0, PLAYER_DELTA_Z_PER_TICK));
+  } else {
+    player->SetDucking(DuckDir::NONE);
+  }
+  if (ImGui::GetIO().KeysDown[GLFW_KEY_LEFT_SHIFT] ||
+      ImGui::GetIO().KeysDown[GLFW_KEY_RIGHT_SHIFT]) {
+    if (player->GetDucking() == DuckDir::NONE) {
+      player->SetDucking(DuckDir::RIGHT);
+    }
+  }
+
+  // Update player position based on new velocity.
+  player->SetPosition(player->GetPosition() +
+                      glm::vec3(DELTA_X_PER_TICK, player->GetYVelocity(),
+                                player->GetZVelocity()));
+}
+
+void PlayerUpdater::AnimatePlayer(std::shared_ptr<GameState> game_state) {
+  std::shared_ptr<Player> player = game_state->GetPlayer();
+  std::shared_ptr<PhysicalObject> rear_wheel = player->GetRearWheel();
+  std::shared_ptr<PhysicalObject> front_wheel = player->GetFrontWheel();
+  Player::Animation current_animation = player->GetAnimation();
+
+  // Rotate the wheels
+  if (current_animation & Player::ANIMATION_WHEELSPIN_BIT) {
+    player->SetWheelRotationSpeed(WHEEL_ROTATION_PER_TICK);
+  } else if (current_animation & Player::ANIMATION_WHEELSPIN_SLOW_BIT) {
+    player->SetWheelRotationSpeed(player->GetWheelRotationSpeed() * 0.98);
+  } else {
+    player->SetWheelRotationSpeed(0);
+  }
+  player->GetRearWheel()->SetRotationAngle(rear_wheel->GetRotationAngle() +
+                                           player->GetWheelRotationSpeed());
+  player->GetFrontWheel()->SetRotationAngle(front_wheel->GetRotationAngle() +
+                                            player->GetWheelRotationSpeed());
+
+  if (current_animation == Player::Animation::FAILURE) {
+    // go crazy
+    player->SetRotationAngle(player->GetRotationAngle() + 0.3f);
+  }
+
+  if (current_animation & Player::ANIMATION_AERIAL_TILT_BIT) {
+    // set player rotation based on y velocity
+    player->SetRotationAngle(std::atan(player->GetYVelocity() * 4));
```

Comment:
It looks a little jaring when the player lands from a long fall can you try this
```cpp
float vel = std::atan(player->GetYVelocity()*4;
if (vel < 0)
  player->SetRotationAngle(std::min(0.872665, vel);
else
    player->SetRotationAngle(std::max(-0.872665, vel);
```
0.872665 is about 50 degrees y The idea here is that the player should never go past +-50 degrees.